code/modules/overmap/ships/computers/engine_control.dm

```diff
@@ -0,0 +1,94 @@
+//Engine control and monitoring console
+
+/obj/machinery/computer/ship/engines
+	name = "engine control console"
+	icon_screen = "enginecontrol"
+	var/display_state = "status"
+
+/obj/machinery/computer/ship/engines/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	if(!linked)
+		display_reconnect_dialog(user, "ship control systems")
+		return
+
+	var/data[0]
+	data["state"] = display_state	
+	data["global_state"] = linked.engines_state
+	data["global_limit"] = round(linked.thrust_limit*100)
+	var/total_thrust = 0
+
+	var/list/enginfo[0]
+	for(var/datum/ship_engine/E in linked.engines)
+		var/list/rdata[0]
+		rdata["eng_type"] = E.name
+		rdata["eng_on"] = E.is_on()
+		rdata["eng_thrust"] = E.get_thrust()
+		rdata["eng_thrust_limiter"] = round(E.get_thrust_limit()*100)
+		rdata["eng_status"] = E.get_status()
+		rdata["eng_reference"] = "\ref[E]"
+		total_thrust += E.get_thrust()
+		enginfo.Add(list(rdata))
+
+	data["engines_info"] = enginfo
+	data["total_thrust"] = total_thrust
+
+	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		ui = new(user, src, ui_key, "engines_control.tmpl", "[linked.name] Engines Control", 390, 530)
+		ui.set_initial_data(data)
+		ui.open()
+		ui.set_auto_update(1)
+
+/obj/machinery/computer/ship/engines/Topic(href, href_list)
+	if(..())
+		return
```

Comment:
ruh roh. this is bad. Shouldn't this be `return TOPIC_NO_ACTION` or just, `. = ..(); return`? Cuz right now we're returning null and losing data and potentially causing a security shenanigans.

Reply:
`TOPIC_HANDLED` is apparently what you're supposed to do here?